### PR TITLE
[Carry #3205] libct/cg: add CFS bandwidth burst for CPU

### PR DESCRIPTION
**Pull request by Zheaoli**
Carry #3205

### CODE REVIEW

**Comment by Zheaoli** [comment]
Seems the systemd is not support the CPU burst feature yet. https://github.com/systemd/systemd/issues/26658

So should we keep pushing this PR or hold on it utill the feature has been supported in systemd @AkihiroSuda

**Comment by AkihiroSuda** [comment]
> Seems the systemd is not support the CPU burst feature yet. [systemd/systemd#26658](https://github.com/systemd/systemd/issues/26658)
> 
> So should we keep pushing this PR or hold on it utill the feature has been supported in systemd @AkihiroSuda

You can just let fs2 handle it
https://github.com/opencontainers/runc/blob/a187c84e42015402263167a2c1007b5c62d494f3/libcontainer/cgroups/systemd/v2.go#L168-L170

**Comment by Zheaoli** [comment]
> > Seems the systemd is not support the CPU burst feature yet. [systemd/systemd#26658](https://github.com/systemd/systemd/issues/26658)
> > So should we keep pushing this PR or hold on it utill the feature has been supported in systemd @AkihiroSuda
> 
> You can just let fs2 handle it
> 
> https://github.com/opencontainers/runc/blob/a187c84e42015402263167a2c1007b5c62d494f3/libcontainer/cgroups/systemd/v2.go#L168-L170

Got it

**Comment by Zheaoli** [comment]
> > Seems the systemd is not support the CPU burst feature yet. [systemd/systemd#26658](https://github.com/systemd/systemd/issues/26658)
> > So should we keep pushing this PR or hold on it utill the feature has been supported in systemd @AkihiroSuda
> 
> You can just let fs2 handle it
> 
> https://github.com/opencontainers/runc/blob/a187c84e42015402263167a2c1007b5c62d494f3/libcontainer/cgroups/systemd/v2.go#L168-L170

But the burst feature supports both cv1 and cv2. hmmm I think maybe I can limit it should be only being used in non-systemd driver?

**Comment by Zheaoli** [comment]
@AkihiroSuda Hi, I have written the test, but the rootless test failed. I have no idea what happened. would you mind take a look at for me ?

**Comment by Zheaoli** [comment]
https://github.com/opencontainers/runc/actions/runs/4462135582/jobs/7836459757?pr=3749
https://github.com/opencontainers/runc/actions/runs/4462008272/jobs/7836249615?pr=3749

It seems strange why `no directory specified for cpu.cfs_burst_us` has been raised but not for other values like period and quota(both of them shared the same path

**Comment by Zheaoli** [comment]
fine, I figure out the reason, all ci has been passed

**Comment by Zheaoli** [comment]
https://cirrus-ci.com/task/5913443228712960
https://cirrus-ci.com/task/4887318499164160
https://cirrus-ci.com/task/4605843522453504

Should I skip the systemd test for burst? cc @AkihiroSuda

**Comment by Zheaoli** [comment]
@AkihiroSuda ping(would you mind to give me some advise about test failed in Centos?

**Comment by AkihiroSuda** [comment]
CI failing

**Comment by Zheaoli** [comment]
@AkihiroSuda It seems like there may be some compatibility issues between CPU Burst and Centos7 and Centos8, which could be causing the CI to fail.  Would your mind to give me some suggestion about how to pass all the CI?

**Comment by AkihiroSuda** [comment]
> It seems like there may be some compatibility issues between CPU Burst and Centos7 and Centos8, which could be causing the CI to fail. Would your mind to give me some suggestion about how to pass all the CI?

I guess you can just check existence of `cpu.cfs_burst_us` and skip the test when it does not exist?

**Comment by Zheaoli** [comment]
> > It seems like there may be some compatibility issues between CPU Burst and Centos7 and Centos8, which could be causing the CI to fail. Would your mind to give me some suggestion about how to pass all the CI?
> 
> I guess you can just check existence of `cpu.cfs_burst_us` and skip the test when it does not exist?

Seems good, I'll make a try! Thanks a lot!

**Comment by Zheaoli** [comment]
@AkihiroSuda Seems all CI has been passed except lint-extra test.

Some warning message has been raised from this step:" var-naming: struct field CpuBurst should be CPUBurst". But I think `CpuBurst` is better than `CPUBurst` because we should keep consistency with other fields?

File: libcontainer/configs/cgroup_linux.go [hunk 1]
```diff
@@ -69,6 +69,9 @@ type Resources struct {
 	// CPU hardcap limit (in usecs). Allowed cpu time in a given period.
 	CpuQuota int64 `json:"cpu_quota"`
 
+	// CPU hardcap burst limit (in usecs). Allowed accumulated cpu time additionally for burst in a given period.
+	CpuBurst uint64 `json:"cpu_burst"`
```
Comment (AkihiroSuda) [inline libcontainer/configs/cgroup_linux.go:73]:
```suggestion
	CpuBurst uint64 `json:"cpu_burst"` //nolint:revive
```

Comment (Zheaoli) [reply libcontainer/configs/cgroup_linux.go]: emmm, it's weird; it seems not need to add an extra flag to other filed. = =

**Review by AkihiroSuda — APPROVED** [review]

**Comment by Zheaoli** [comment]
@kolyshkin Hello. Would you mind looking at this PR if you have time? Thanks!

**Comment by AkihiroSuda** [comment]
Fedora CI is failing, rebasing may solve the issue.

**Comment by AkihiroSuda** [comment]
Please update https://github.com/opencontainers/runc/blob/main/docs/spec-conformance.md too

File: libcontainer/configs/cgroup_linux.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainer/configs/cgroup_linux.go]: This is because lint-extra is a job which runs more strict linters on new code only. But yes, I agree that if other fields are named `Cpu*`, we should use `Cpu`, too.

**Comment by Zheaoli** [comment]
@AkihiroSuda Thanks for your tips, and sorry for replying late. 

I have merged the newest code from the main branch and updated the docs. But there is still some issues:

1. the CI on Centos-Stream-9 has failed. The message is `Command "grep -q oom events.log" failed 30 times. Output:`. I think this issue is not relevant to this PR. Would you mind rerunning this step for me?
2. The spell check has failed. I'm sure if I can skip the check or not.

**Review by AkihiroSuda — APPROVED** [review]

**Comment by Zheaoli** [comment]
@kolyshkin ping~

**Comment by lifubang** [comment]
@Zheaoli I think you should use `git rebase main` instead of `git merge main`.

**Comment by Zheaoli** [comment]
> @Zheaoli I think you should use `git rebase main` instead of `git merge main`.

copy that, I will squash commit later

File: libcontainer/cgroups/fs/cpu.go [hunk 2]
```diff
@@ -84,6 +84,28 @@ func (s *CpuGroup) Set(path string, r *configs.Resources) error {
...
+		if err := cgroups.WriteFile(path, "cpu.cfs_burst_us", burst); err != nil {
+			// Sometimes when the burst to be set is larger
+			// than the current one, it is rejected by the kernel
+			// (EINVAL) as old_quota/new_burst exceeds the parent
+			// cgroup quota limit. If this happens and the quota is
+			// going to be set, ignore the error for now and retry
+			// after setting the quota.
+			if !errors.Is(err, unix.ENOENT) {
+				// this is a special trick for burst feature, the current systemd and low version of kernel will not support it.
+				// So, an `no such file or directory` error would be raised, and we can ignore it .
```
Comment (lifubang) [inline libcontainer/cgroups/fs/cpu.go:100]: I think these two lines comments is for `unix.ENOENT`, you should put the comments in the correct place, so please exchange the comments of `unix.ENOENT` and `unix.EINVAL`.

File: libcontainer/cgroups/fs/cpu.go [hunk 3]
```diff
@@ -93,6 +115,11 @@ func (s *CpuGroup) Set(path string, r *configs.Resources) error {
 				return err
 			}
 		}
+		if burst != "" {
+			if err := cgroups.WriteFile(path, "cpu.cfs_burst_us", burst); err != nil {
+				return err
```
Comment (lifubang) [inline libcontainer/cgroups/fs/cpu.go:120]: Should we ignore the error of `unix.ENOENT` here too?

File: libcontainer/cgroups/fs2/cpu.go [hunk 4]
```diff
@@ -32,6 +35,23 @@ func setCpu(dirPath string, r *configs.Resources) error {
...
+	if r.CpuBurst != 0 {
+		burst = strconv.FormatUint(r.CpuBurst, 10)
+		if err := cgroups.WriteFile(dirPath, "cpu.max.burst", burst); err != nil {
+			// Sometimes when the burst to be set is larger
+			// than the current one, it is rejected by the kernel
+			// (EINVAL) as old_quota/new_burst exceeds the parent
+			// cgroup quota limit. If this happens and the quota is
+			// going to be set, ignore the error for now and retry
+			// after setting the quota.
+			if !errors.Is(err, unix.EINVAL) || r.CpuQuota == 0 {
```
Comment (lifubang) [inline libcontainer/cgroups/fs2/cpu.go:48]: I don't know whether there is a case of `unix.ENOENT` in `cgroupv2` or not? I think there is a high probability that there is no this type case. But maybe I'm wrong.

**Review by lifubang — CHANGES_REQUESTED** [review]

File: libcontainer/cgroups/fs2/cpu.go (hunk 4 above)
Comment (Zheaoli) [reply libcontainer/cgroups/fs2/cpu.go:48]: I think it's good keep the same code style between cgroupv1 and v2

Comment (lifubang) [reply libcontainer/cgroups/fs2/cpu.go:48]: So we need add the same logic for  'unix.ENOENT' like v1?

Comment (Zheaoli) [reply libcontainer/cgroups/fs2/cpu.go:48]: Yep, I think this would be a good idea

**Comment by lifubang** [comment]
👍 Please use ‘git rebase main’ to catch up with the latest change, we should not have a commit named ‘merge main’. Thanks.

**Comment by Zheaoli** [comment]
@lifubang Thanks for the review, all things done

File: update.go [hunk 5]
```diff
@@ -298,6 +305,12 @@ other options are ignored.
 			}
 		}
 
+		b := *r.CPU.Burst
+		if config.Cgroups.Resources.CpuPeriod == 0 && b != 0 {
```
Comment (lifubang) [inline update.go:309]: @Zheaoli Sorry for delayed so much time. Here I understand you mean that `cpu-burst` depends on `cpu-period`? But I think may be we should depends on `cpu-quota`, or maybe we should remove this dependent check? Let it check by the kernel?

Comment (Zheaoli) [reply update.go:309]: To be honest, I think it would be great to validate some base rule at the entry of the CLI like https://github.com/opencontainers/runc/blob/main/update.go#L318-L325

If we put the validation process in the kernel, I think the error message will confuse people.

Comment (lifubang) [reply update.go:309]: > I think the error message will confuse people.

But there is no such check when running a container. There would be different results for ‘run’ and ‘update’ if we only provide ’cpu-burst’.

File: libcontainer/cgroups/fs/cpu.go [hunk 6]
```diff
@@ -84,6 +84,28 @@ func (s *CpuGroup) Set(path string, r *configs.Resources) error {
 			period = ""
 		}
 	}
+
+	var burst string
+	if r.CpuBurst != 0 {
```
Comment (lifubang) [inline libcontainer/cgroups/fs/cpu.go:89]: If we want to clear `cpu-burst`, the value should be 0. So we should change `r.CpuBurst` to a pointer and to check whether this is nil or not.

File: libcontainer/configs/cgroup_linux.go [hunk 7]
```diff
@@ -69,6 +69,9 @@ type Resources struct {
 	// CPU hardcap limit (in usecs). Allowed cpu time in a given period.
 	CpuQuota int64 `json:"cpu_quota"`
 
+	// CPU hardcap burst limit (in usecs). Allowed accumulated cpu time additionally for burst in a given period.
+	CpuBurst uint64 `json:"cpu_burst"` //nolint:revive
```
Comment (lifubang) [inline libcontainer/configs/cgroup_linux.go:73]: Please change to `CpuBurst *unit64`

**Review by lifubang — CHANGES_REQUESTED** [review]

**Comment by lifubang** [comment]
I'm so sorry to let you spend more time on this. We need to cover as many test cases as possible. ☕

File: libcontainer/configs/cgroup_linux.go (hunk 7 above)
Comment (Zheaoli) [reply libcontainer/configs/cgroup_linux.go:73]: I think the filed type should be as same as other fields like `CpuQuota`

**Comment by Zheaoli** [comment]
> We need to cover as many test cases as possible. ☕

I think we already have enough test cases here. May I ask what's kind of the corner case in your thought not covered by current test code?

File: libcontainer/configs/cgroup_linux.go (hunk 7 above)
Comment (lifubang) [reply libcontainer/configs/cgroup_linux.go:73]: No, For `CpuQuota`, `-` means `no limit`, but for `CpuBurst`, `0` means `no config`.

File: libcontainer/cgroups/fs/cpu.go (hunk 6 above)
Comment (Zheaoli) [reply libcontainer/cgroups/fs/cpu.go:89]: I think splitting the nil check and zero-value check process is unnecessary here.

In my opinion, the nil value(the user doesn't input the burst value) and the zero-value(the user input burst as zero) means the same thing for us which is we should not write the burst value to the cgroup file

**Comment by lifubang** [comment]
> May I ask what's kind of the corner case in your thought not covered by current test code?

For example, how to clear `cpu-burst` value in `update` after the container started with a `cpu-burst` value?

File: libcontainer/cgroups/fs/cpu.go (hunk 6 above)
Comment (lifubang) [reply libcontainer/cgroups/fs/cpu.go:89]: For example, the container `test` started with the config:
```
cat config.json | jq .linux.resources.cpu
{
  "burst": 1000
}
```

After some time, we want to clear burst value, we may run:
```
runc update --cpu-burst 0 test
```

File: libcontainer/configs/cgroup_linux.go (hunk 7 above)
Comment (lifubang) [reply libcontainer/configs/cgroup_linux.go:73]: I think `0` is a legal value which should be allowed to write to cgroup.

Comment (Zheaoli) [reply libcontainer/configs/cgroup_linux.go:73]: SGTM

**Review by lifubang — APPROVED** [review]
LGTM

**Comment by lifubang** [comment]
Left some further works:
1. No unit test for cpu controller in cgroupv2;
2. There are some different results for cgroup v1 and v2, for example:
```
runc update --cpu-burst 30001 --cpu-quota 30000 test
```
v1: fail, v2: succ
Because for v2, the kernel write seconds to `cpu.max.burst`, but not milliseconds.
3. There is no integration test for `cpu-burst` when create/run a container.

We can do these works in further PRs.

**Comment by lifubang** [comment]
@AkihiroSuda PTAL. I think this could be merged.

**Review by AkihiroSuda — APPROVED** [review]
